crates/handlers/src/compat/mod.rs

```diff
@@ -27,3 +37,112 @@ impl IntoResponse for MatrixError {
+                status: StatusCode::BAD_REQUEST,
+            },
+
+            Self::BytesRejection(BytesRejection::FailedToBufferBody(
+                FailedToBufferBody::LengthLimitError(_),
+            )) => MatrixError {
+                errcode: "M_TOO_LARGE",
+                error: "Request body too large",
+                status: StatusCode::PAYLOAD_TOO_LARGE,
+            },
+
+            Self::BytesRejection(BytesRejection::FailedToBufferBody(
+                FailedToBufferBody::UnknownBodyError(_),
+            )) => MatrixError {
+                errcode: "M_UNKNOWN",
+                error: "Failed to read request body",
+                status: StatusCode::BAD_REQUEST,
+            },
+
+            Self::BytesRejection(_) => MatrixError {
+                errcode: "M_UNKNOWN",
+                error: "Unknown error while reading request body",
+                status: StatusCode::BAD_REQUEST,
+            },
+
+            Self::NotJson(_) => MatrixError {
+                errcode: "M_NOT_JSON",
+                error: "Body is not a valid JSON document",
+                status: StatusCode::BAD_REQUEST,
+            },
+
+            Self::BadJson(_) => MatrixError {
+                errcode: "M_BAD_JSON",
+                error: "JSON fields are not valid",
+                status: StatusCode::BAD_REQUEST,
+            },
+        };
+
+        (SentryEventID::from(event_id), response).into_response()
+    }
+}
+
+impl<T, S> axum::extract::FromRequest<S> for MatrixJsonBody<T>
+where
+    T: DeserializeOwned,
+    S: Send + Sync,
+{
+    type Rejection = MatrixJsonBodyRejection;
+
+    async fn from_request(req: Request, state: &S) -> Result<Self, Self::Rejection> {
+        // Matrix spec says it's optional to send a Content-Type header, so we
+        // only check it if it's present
+        if let Some(content_type) = req.headers().get(header::CONTENT_TYPE) {
+            let Ok(content_type) = content_type.to_str() else {
+                return Err(MatrixJsonBodyRejection::InvalidContentType);
+            };
+
+            let Ok(mime) = content_type.parse::<mime::Mime>() else {
+                return Err(MatrixJsonBodyRejection::InvalidContentType);
+            };
+
+            let is_json_content_type = mime.type_() == "application"
+                && (mime.subtype() == "json" || mime.suffix().is_some_and(|name| name == "json"));
+
+            if !is_json_content_type {
+                return Err(MatrixJsonBodyRejection::ContentTypeNotJson(mime));
+            }
+        }
+
+        let bytes = Bytes::from_request(req, state).await?;
+
+        // We first parse it as a JSON value so that we can distinguish between
+        // invalid JSON documents and invalid requests
+        let value: serde_json::Value =
+            serde_json::from_slice(&bytes).map_err(MatrixJsonBodyRejection::NotJson)?;
+        let value = serde_json::from_value(value).map_err(MatrixJsonBodyRejection::BadJson)?;
```

Comment:
I guess these requests are tiny so this is fine. But if they were big, I'd rather avoid this and you could use `.is_syntax()` on the serde_json Error.

Reply:
You're right, better do that. I used `.is_data()` instead

7f4e97535